compiler/dyno/tools/chpldoc/arg-helpers.cpp

```diff
@@ -0,0 +1,239 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "arg-helpers.h"
+#include "chpl/util/filesystem.h"
+
+bool developer = false;
+
+
+bool startsWith(const char* str, const char* prefix) {
+  return (0 == strncmp(str, prefix, strlen(prefix)));
+}
+
+void clean_exit(int status) {
+  exit(status);
+}
+
+void setupError(std::string msg) {
+  printf("setupError %s", msg.c_str());
+}
+
+
+uint64_t hexStr2uint64(const char* str, bool userSupplied,
+                       const char* filename, int line) {
+  if (!str) {
+    INT_FATAL("NULL string passed to hexStrToUint64()");
+  }
+  int len = strlen(str);
+  if (len < 3 || str[0] != '0' || (str[1] != 'x' && str[1] != 'X')) {
+    INT_FATAL("Illegal string passed to hexStrToUint64()");
+  }
+  /* Remove leading 0s */
+  int startPos = 2;
+  while (str[startPos] == '0' && startPos < len-1) {
+    startPos++;
+  }
+
+  if (strlen(str+startPos) > 16) {
+    if (userSupplied) {
+      //astlocT astloc(line, filename);
+      USR_FATAL("Integer literal overflow: '" + std::string(str) + "' is too big "
+                "for type 'uint64'");
+    } else {
+      INT_FATAL("Integer literal overflow: '" + std::string(str) + "' is too big "
+                "for type 'uint64'");
+    }
+  }
+
+  uint64_t val;
+  int numitems = sscanf(str+2, "%" SCNx64, &val);
+  if (numitems != 1) {
+    INT_FATAL("Illegal string passed to hexStrToUint64");
+  }
+  return val;
+}
+
+// // would just use realpath, but it is not supported on all platforms.
+static
+char* chplRealPath(const char* path)
+{
+  // We would really rather use
+  // char* got = realpath(path, NULL);
+  // but that doesn't work on some Mac OS X versions.
+  char* buf = (char*) malloc(1024);
+  char* got = realpath(path, buf);
+  char* ret = NULL;
+  if( got ) ret = strdup(got);
+  free(buf);
+  return ret;
+}
+
+
+// Returns a "real path" to the file in the directory,
+// or NULL if the file did not exist.
+// The return value must be freed by the caller.
+// We try to use realpath but might give up.
+static
+char* dirHasFile(const char *dir, const char *file)
+{
+  struct stat stats;
+  int len = strlen(dir) + strlen(file) + 2;
+  char* tmp = (char*) malloc(len);
+  char* real;
+
+  if( ! tmp ) INT_FATAL("no memory");
+
+  snprintf(tmp, len, "%s/%s", dir, file);
+  real = chplRealPath(tmp);
+  if( real == NULL ) {
+    // realpath not working on this system,
+    // just use tmp.
+    real = tmp;
+  } else {
+    free(tmp);
+  }
+
+  if( stat(real, &stats) != 0) {
+    free(real);
+    real = NULL;
+  }
+
+  return real;
+}
+
+
+// Find the path to the running program
+// (or return NULL if we couldn't figure it out).
+// The return value must be freed by the caller.
+char* findProgramPath(const char *argv0)
+{
```

Comment:
There is an LLVM support library call you could use to do this same thing.

Reply:
It is `getMainExecutable`

Reply:
I implemented that in the next PR

Reply:
https://github.com/chapel-lang/chapel/pull/20558 is the draft of the PR that changes `findProgramPath` to use `getMainExecutable`